src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,320 @@
+
+            return false;
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            ISymbol symbol = context.Symbol;
+            if (symbol == null)
+            {
+                return;
+            }
+
+            // Don't report diagnostics on types that are marked as Preview
+            if (HasPreviewAttribute(symbol, requiresPreviewFeaturesSymbols))
+            {
+                return;
+            }
+
+            bool reportDiagnostic = false;
+
+            // Look for attributes on base classes that this type may extend
+            if (symbol is ITypeSymbol typeSymbol)
+            {
+                reportDiagnostic = TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType);
+            }
+
+            if (!reportDiagnostic)
+            {
+                if (symbol is IMethodSymbol || symbol is IPropertySymbol)
+                {
+                    reportDiagnostic = PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType);
+                }
+            }
+
+            if (reportDiagnostic)
+            {
+                if (requiresPreviewFeaturesSymbolsToUsageType.TryGetValue(symbol, out PreviewFeatureUsageType usageType))
+                {
+                    if (usageType == PreviewFeatureUsageType.StaticAbstract)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(StaticAbstractIsPreviewFeatureRule));
+                    }
+                    else if (usageType == PreviewFeatureUsageType.EmptyInterface)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(ImplementsEmptyPreviewInterfaceRule, symbol.Name));
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(symbol.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+                }
+            }
+        }
+
+        private static void BuildSymbolInformationFromOperations(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            IOperation operation = context.Operation;
+            if (OperationUsesPreviewFeatures(operation, requiresPreviewFeaturesSymbols, out ISymbol? symbol))
+            {
+                if (operation is ICatchClauseOperation catchClauseOperation)
+                {
+                    operation = catchClauseOperation.ExceptionDeclarationOrExpression;
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol!.Name));
```

Comment:
Instead of the suppression, shouldn't adding `[NotNullWhen(true)]` to the out parameter of `OperationUsesPreviewFeatures` fix that?

Reply:
Nice catch. Fixed.